### PR TITLE
feat: add GracefulContext

### PR DESCRIPTION
**Pull request by alnr**
There's currently no convenient way to shut down a server from within the same process other than sending SIGINT/SIGTERM to itself. This function adds that capability.

### CODE REVIEW

File: http_graceful_test.go [hunk 1]
```diff
@@ -72,6 +72,31 @@ func TestGraceful(t *testing.T) {
...
+		}()
+
+		// Kill the server after 1s
+		go func() {
+			time.Sleep(1 * time.Second)
+			cancel()
+		}()
+
+		_, err := http.Get("http://localhost:54933/")
+		require.Error(t, err)
```
Comment (zepatrik) [inline http_graceful_test.go:111]: Why do we expect to get an error here? What error do we get?

Comment (alnr) [reply http_graceful_test.go:111]: I've bumped the dependencies and improved the test a little bit.

Comment (zepatrik) [reply http_graceful_test.go:111]: Ah I see, the server has a delay of 10s, but we cancel it after 1s. This way it can never complete this request.

**Review by zepatrik — APPROVED** [review]
Looks great :+1:

**Review by hperl — APPROVED** [review]
Nice, LGTM!

**Pull request merged**